pkg/operators/kubenameresolver/kubenameresolver.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package kubenameresolver provides an operator that enriches events by looking
+// up the pod name and namespace and enriches it with its ip information. It is
+// currently used by the following gadgets:
+// - trace network
+package kubenameresolver
+
+import (
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/common"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators/kubemanager"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	OperatorName = "KubeNameResolver"
+)
+
+type KubeNameResolverInterface interface {
+	SetLocalPodDetails(owner, hostIP, podIP string, labels map[string]string)
+}
+
+type KubeNameResolver struct {
+	k8sInventory *common.K8sInventoryCache
+}
+
+func (k *KubeNameResolver) Name() string {
+	return OperatorName
+}
+
+func (k *KubeNameResolver) Description() string {
+	return "KubeNameResolver resolves pod name/namespace to IP addresses..."
+}
+
+func (k *KubeNameResolver) GlobalParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (k *KubeNameResolver) ParamDescs() params.ParamDescs {
+	return nil
+}
+
+func (k *KubeNameResolver) Dependencies() []string {
+	return []string{kubemanager.OperatorName}
+}
+
+func (k *KubeNameResolver) CanOperateOn(gadget gadgets.GadgetDesc) bool {
+	km := kubemanager.KubeManager{}
+	if !km.CanOperateOn(gadget) {
+		return false
+	}
+	_, hasNameResolverInterface := gadget.EventPrototype().(KubeNameResolverInterface)
+	return hasNameResolverInterface
```

Comment:
I know it was already like that in the KubeIPResolver, but I can't remember why we need to ensure that the `KubeManager` can operate on the gadget instead of simply doing:

```suggestion
	_, hasContainerInfoGetters := gadget.EventPrototype().(operators.ContainerInfoGetters)
	_, hasNameResolverInterface := gadget.EventPrototype().(KubeNameResolverInterface)
	return hasNameResolverInterface && hasContainerInfoGetters
```

Reply:
Reason is that we need `KubeManager` to work at all. That is the operator which sets the information which can be retrieved by `ContainerInfoGetters`. This way we always stay aligned with `KubeManager`

Reply:
> Afaik, LocalManager will soon also be able to enrich Kubernetes metadata as well, right?
>
Actually, the LocalManager is already enriching with the k8s metadata. The problem is that the container name is not the one from k8s but from the container runtime. https://github.com/inspektor-gadget/inspektor-gadget/pull/1702 uses the correct container name (btw, reviews are welcome there). 

> Importing `KubeManager` here directly installs it as operator
>
It's already done in the gadget-service, so it shouldn't be necessary here. Or did I miss/misunderstand something?

> That is the operator which sets the information which can be retrieved by `ContainerInfoGetters`. This way we always stay aligned with `KubeManager`
>
IMO, this is something the `KubeNameResolverInstance` shouldn't care about. I mean, in practice, it just requires an event from which it can extract the namespace and the pod. It doesn't care how such information was added to the event. I think we `CanOperateOn` should be:

```go
func (k *KubeNameResolver) CanOperateOn(gadget gadgets.GadgetDesc) bool {
	_, hasContainerInfoGetters := gadget.EventPrototype().(operators.ContainerInfoGetters)
	_, hasNameResolverInterface := gadget.EventPrototype().(KubeNameResolverInterface)
	return hasNameResolverInterface && hasContainerInfoGetters
}
```

At most, we can do this:

```go
func (m *KubeNameResolver) enrich(ev any) {
	additionalInfo, _ := ev.(KubeNameResolverInterface)
	containerInfo, _ := ev.(operators.ContainerInfoGetters)

	if containerInfo.GetNamespace() == "" || containerInfo.GetPod() == "" {
		log.Debugf("Impossible to enrich event with empty namespace/pod info")
		return
	}

	[...]
	}
```

And the `Dependencies` shouldn't be a specific operator but something like "I need an operator that enriches the event with this data" ... No idea how to do it, though 🤣.

Anyway, we can manage this in another PR.

Reply:
> > That is the operator which sets the information which can be retrieved by `ContainerInfoGetters`. This way we always stay aligned with `KubeManager`
> 
> IMO, this is something the `KubeNameResolverInstance` shouldn't care about. I mean, in practice, it just requires an event from which it can extract the namespace and the pod. It doesn't care how such information was added to the event. I think we `CanOperateOn` should be:
> 

I think you are right here. See below

> And the `Dependencies` shouldn't be a specific operator but something like "I need an operator that enriches the event with this data" ... No idea how to do it, though rofl.

Guess this is the hard part :D. Probably can't be done with the current mechanisms